<div style="text-align:center; padding:20px; border:2px solid #333; border-radius:10px; background:#f5f5f5;">
<h1 style="color:#2ca02c;">Project 1:</h1>
<h1 style="color:#1f77b4;">Data Cleaning & Preparation</h1>
<h3>Developed by : HAJAR ELKAIDI</h3>



**In this project, I worked on cleaning and organizing a raw dataset by dealing with missing values, removing duplicates, and correcting inconsistencies. The aim was to enhance the quality of the data and prepare a trustworthy dataset that can be used for analysis and decision-making.**

<span style="color:#ff0000; font-weight:bold;">1- IMPORTS</span>

In [2]:
import pandas as pd

<span style="color:#ff0000; font-weight:bold;">2- DATA UNDERSTANDING :</span>

<p style="color:#2ca02c;"><b>a)- Data loading :</b></p>

In [3]:
data = pd.read_excel("Dataset_for_Data_Analytics.xlsx")

<p style="color:#2ca02c;"><b>b)- Columns & Data Types:</b></p>

In [4]:
data.head(10)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04
5,ORD200005,2023-10-23,C37249,Phone,2,245.86,934 Main St,Credit Card,Shipped,TRK72976927,4,SAVE10,Instagram,491.72
6,ORD200006,2025-06-17,C83492,Laptop,1,664.42,986 Main St,Gift Card,Returned,TRK96417362,6,SAVE10,Facebook,664.42
7,ORD200007,2023-05-12,C41460,Monitor,5,149.55,706 Main St,Cash,Shipped,TRK78809193,9,FREESHIP,Facebook,747.75
8,ORD200008,2025-04-02,C26817,Phone,2,134.28,904 Main St,Gift Card,Cancelled,TRK61042692,2,NaN,Email,268.56
9,ORD200009,2023-11-21,C31946,Desk,4,509.38,102 Main St,Credit Card,Shipped,TRK33478363,6,SAVE10,Google,2037.52


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The output of `data.head(10)` gave a quick first look at the dataset and how it is structured. It showed the first ten rows, which included important details about customer orders such as order information, products, quantities, payment methods, order status, and total prices. This initial view helped us better understand the data and prepare it for further analysis.


In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumber   1200 non-null   object        
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    object        
 12  ReferralSource   1200 non-null   object        
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(

<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The data.info() function was used to take a closer look at the structure of the dataset. It shows that we have 1,200 rows and 14 columns, with different types of data like integers, floats, text (objects), and datetime values. This step was important because it helped us understand how the data is organized and check if everything is in the right format before moving on to cleaning and analysis.

<span style="color:#FF0000; font-weight:bold;">3- DATA QUALITY ASSESSEMENT :</span>

<p style="color:#2ca02c;"><b>a)- Check Missing Values :</b></p>

In [6]:
        # Number of missing values: 
missing_values = data.isnull().sum()
print("** Missing values before cleaning are: **\n",missing_values)
        # Percentage of missing values:
per_missing_values = data.isnull().mean() * 100
print("\n** The percentages of missing values are: **\n",per_missing_values)

** Missing values before cleaning are: **
 OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

** The percentages of missing values are: **
 OrderID             0.00
Date                0.00
CustomerID          0.00
Product             0.00
Quantity            0.00
UnitPrice           0.00
ShippingAddress     0.00
PaymentMethod       0.00
OrderStatus         0.00
TrackingNumber      0.00
ItemsInCart         0.00
CouponCode         25.75
ReferralSource      0.00
TotalPrice          0.00
dtype: float64


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The analysis of missing values showed that most of the dataset is complete, with no missing data in the majority of columns. However, the `CouponCode` column had 309 missing values, which is about 25.75% of the dataset. This suggests that a significant number of orders were made without using a coupon code, while all the other fields are fully filled.


<p style="color:#2ca02c;"><b>b)- Check Duplicates :</b></p>

In [7]:
        # Search duplicate rows:
print("Duplicate rows:", data.duplicated().sum())
        # Search duplicate Order IDs:
print("Duplicate Order IDs:", data["OrderID"].duplicated().sum())

Duplicate rows: 0
Duplicate Order IDs: 0


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The duplicate check showed that there are no repeated rows in the dataset, and no duplicate Order IDs were found. This means each order is unique, which helps ensure the reliability of the data for further analysis.


<p style="color:#2ca02c;"><b>c)- Check Unique Values :</b></p>

In [8]:
categorical_cols = ['Product', 'PaymentMethod', 'OrderStatus', 'CouponCode', 'ReferralSource']

for col in categorical_cols:
    print(f"\n{col}:")
    print(data[col].unique())


Product:
['Monitor' 'Phone' 'Tablet' 'Chair' 'Printer' 'Laptop' 'Desk']

PaymentMethod:
['Debit Card' 'Online' 'Credit Card' 'Gift Card' 'Cash']

OrderStatus:
['Shipped' 'Cancelled' 'Returned' 'Delivered' 'Pending']

CouponCode:
['SAVE10' 'FREESHIP' nan 'WINTER15']

ReferralSource:
['Instagram' 'Referral' 'Email' 'Facebook' 'Google']


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The categorical columns were reviewed to check the unique values and make sure there were no inconsistencies. Overall, the categories were clean and consistent, with no spelling mistakes or formatting issues detected. The only missing values appeared in the `CouponCode` column, which simply indicates orders where no discount coupon was used.


<p style="color:#2ca02c;"><b>d)- Check Business Rules :</b></p>

In [9]:
# Quantity:
invalid_quantity = data[data['Quantity'] <= 0]
print("Invalid Quantity:", invalid_quantity.shape[0])

# UnitPrice:
invalid_price = data[data['UnitPrice'] <= 0]
print("Invalid UnitPrice:", invalid_price.shape[0])

#ItemsInCart:
invalid_cart = data[data['ItemsInCart'] < 1]
print("Invalid ItemsInCart:", invalid_cart.shape[0])

#TotalPrice:
invalid_total = data[
    data['TotalPrice'].round(2) !=
    (data['Quantity'] * data['UnitPrice']).round(2)]
print("Invalid TotalPrice:", invalid_total.shape[0])

Invalid Quantity: 0
Invalid UnitPrice: 0
Invalid ItemsInCart: 0
Invalid TotalPrice: 0


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
All the business rules were checked and everything was valid. The values in `Quantity`, `UnitPrice`, and `ItemsInCart` were all correct, and `TotalPrice` properly matched the calculation of Quantity × UnitPrice, with only minor differences due to decimal precision. No violations or inconsistencies were found in the data.


<p style="color:#2ca02c;"><b>e)- Outliers Detection :</b></p>

In [10]:
numeric_cols = ["Quantity", "UnitPrice", "ItemsInCart", "TotalPrice"]
for col in numeric_cols:
    # Calculate the first quartile (25%)
    Q1 = data[col].quantile(0.25)

    # Calculate the third quartile (75%)
    Q3 = data[col].quantile(0.75)

    # Compute the Interquartile Range (IQR)
    IQR = Q3 - Q1

    # Define the lower and upper bounds for outliers
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    # Filter rows where values are outside the normal range
    outliers = data[
        (data[col] < lower) |
        (data[col] > upper)
    ]
    # Print the number of outliers found in the column
    print(f"\n{col}: {outliers.shape[0]} outliers")
    # Display important details of the detected outliers
    print(outliers[["OrderID", "Product", "Quantity", "UnitPrice", "TotalPrice"]])


Quantity: 0 outliers
Empty DataFrame
Columns: [OrderID, Product, Quantity, UnitPrice, TotalPrice]
Index: []

UnitPrice: 0 outliers
Empty DataFrame
Columns: [OrderID, Product, Quantity, UnitPrice, TotalPrice]
Index: []

ItemsInCart: 0 outliers
Empty DataFrame
Columns: [OrderID, Product, Quantity, UnitPrice, TotalPrice]
Index: []

TotalPrice: 8 outliers
        OrderID  Product  Quantity  UnitPrice  TotalPrice
107   ORD200107  Printer         5     670.75     3353.75
326   ORD200326   Laptop         5     670.48     3352.40
328   ORD200328   Tablet         5     674.04     3370.20
469   ORD200469    Chair         5     676.98     3384.90
632   ORD200632   Laptop         5     678.16     3390.80
789   ORD200789   Tablet         5     691.28     3456.40
1065  ORD201065  Printer         5     666.80     3334.00
1122  ORD201122  Monitor         5     678.19     3390.95


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The outlier analysis showed that `Quantity`, `UnitPrice`, and `ItemsInCart` do not contain any unusual values, which means the data in these columns is consistent. On the other hand, `TotalPrice` has 8 outliers, representing some unusually high transaction amounts. These cases may need further investigation to understand whether they are valid transactions or potential anomalies.


<span style="color:#FF0000; font-weight:bold;">4- DATA CLEANING & PREPARATION :</span>

<p style="color:#2ca02c;"><b>a)- Handling Missing Values :</b></p>

In [15]:
        # Fill missing CouponCode values:
data["CouponCode"] = data["CouponCode"].fillna("NOT USED")
        # Verify missing values:
print("Missing values after cleaning are:\n",data.isnull().sum())

Missing values after cleaning are:
 OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The missing values in the `CouponCode` column most likely mean that no coupon was applied during those purchases. To handle this, we will replace the missing values with "Not Used". This helps us keep the information clear and consistent, without losing any rows from the dataset.


<p style="color:#2ca02c;"><b>b)- Removing Duplicates :</b></p>

<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
Duplicate rows: 0<br>
Duplicate Order IDs: 0<br>
No duplicate records were found, so no rows were removed.

<p style="color:#2ca02c;"><b>c)- Standarizing Text :</b></p>

In [21]:
text_columns = ["PaymentMethod", "OrderStatus", "ReferralSource"]
for col in text_columns:
    data[col] = data[col].str.strip().str.title()
data[text_columns].head()

,PaymentMethod,OrderStatus,ReferralSource
0,Debit Card,Shipped,Instagram
1,Online,Shipped,Referral
2,Credit Card,Cancelled,Email
3,Debit Card,Returned,Facebook
4,Online,Delivered,Email


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The text fields were checked for any formatting issues, spelling mistakes, or unnecessary spaces. The results showed that everything was already clean and consistent, so no changes were needed.


<p style="color:#2ca02c;"><b>d)- Correcting Data Types :</b></p>

<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
The data types were reviewed and found to be correctly assigned for all columns. As a result, no type conversions were necessary.


<p style="color:#2ca02c;"><b>e)- Handling Outliers :</b></p>

<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
Potential outliers were checked using both business rules and statistical analysis. While a few extreme values were identified, they were still considered valid cases, such as large orders or high-value purchases. Because of this, no outliers were removed or changed in order to keep the data accurate and realistic.


<span style="color:#FF0000; font-weight:bold;">5- EXPORT CLEANED DATASET :</span>

In [22]:
data.to_excel("cleaned_dataset.xlsx", index=False)
print("Cleaned dataset exported successfully!")

Cleaned dataset exported successfully!


<span style="color:#1f77b4;"><b><u>Comment:</u></b></span><br>
After finishing the data cleaning and preparation process, the final version of the dataset was exported and saved for further analysis. This makes sure that all the preprocessing steps such as handling missing values, checking duplicates, validating data consistency, and reviewing outliers are properly preserved in a clean and structured format for future use.
